R/aggregate_game_stats_def.R

```diff
@@ -0,0 +1,675 @@
+  dplyr::select(
+    tidyselect::contains("safety_player_")
+  ) %>%
+  tidyr::pivot_longer(cols = c(
+    tidyselect::contains("safety_player_")
+  ),
+  names_to = "desc",
+  names_prefix = "safety_",
+  values_to = "safety_player_id",
+  values_drop_na = TRUE
+  ) %>%
+  dplyr::filter(grepl("_id",desc)) %>%
+  dplyr::mutate(n = 1) %>%
+  tidyr::pivot_wider(
+    names_from = desc,
+    values_from = n,
+    values_fn = sum
+  ) %>%
+  dplyr::select(
+    player_id = safety_player_id,
+    safety = .data$player_id
+  ) %>%
+  dplyr::ungroup()
+
+safety_df_nas <- is.na(safety_df)
+safety_df[safety_df_nas] <- 0
+
+# Penalty stats -----------------------------------------------------------
+
+# get penalty stats
+penalty_df <- data %>%
+  ### the following lines are just to separate defense from offense --> is there a more effitient way???
+  dplyr::filter(!(is.na(penalty_player_id))) %>%
+  dplyr::filter(defteam==penalty_team) %>%
+  ###
+  dplyr::select(
+    tidyselect::contains("penalty_player_")
+  ) %>%
+  tidyr::pivot_longer(cols = c(
+    tidyselect::contains("penalty_player_")
+  ),
+  names_to = "desc",
+  names_prefix = "penalty_",
+  values_to = "penalty_player_id",
+  values_drop_na = TRUE
+  ) %>%
+  dplyr::filter(grepl("_id",desc)) %>%
+  dplyr::mutate(n = 1) %>%
+  tidyr::pivot_wider(
+    names_from = desc,
+    values_from = n,
+    values_fn = sum
+  ) %>%
+  dplyr::select(
+    player_id = penalty_player_id,
+    penalty = .data$player_id
+  ) %>%
+  dplyr::ungroup()
+
+penalty_df_nas <- is.na(penalty_df)
+penalty_df[penalty_df_nas] <- 0
+
+# Touchdown stats -----------------------------------------------------------
+
+# get defensive touchdowns
+touchdown_df <- data %>%
+  dplyr::filter(touchdown==1) %>%
+  dplyr::filter(defteam==td_team) %>%
+  dplyr::group_by(player_id=td_player_id) %>%
+  dplyr::summarise(td = sum(touchdown)) %>%
+  dplyr::ungroup()
+
+touchdown_df_nas <- is.na(touchdown_df)
+touchdown_df[touchdown_df_nas] <- 0
+
+
+# Combine all stats -------------------------------------------------------
+
+# combine all the stats together
+
+player_df <- tackle_df %>%
+  dplyr::full_join(tackle_yds_df, by = c("player_id")) %>%
+  dplyr::full_join(pressure_df, by = c("player_id")) %>%
+  dplyr::full_join(int_df, by = c("player_id")) %>%
+  dplyr::full_join(sack_yds_df, by = c("player_id")) %>%
+  dplyr::full_join(fumble_df_own, by = c("player_id")) %>%
+  dplyr::full_join(fumble_df_opp, by = c("player_id")) %>%
+  dplyr::full_join(fumble_yds_own_df, by = c("player_id")) %>%
+  dplyr::full_join(fumble_yds_opp_df, by = c("player_id")) %>%
+  dplyr::full_join(safety_df, by = c("player_id")) %>%
+  dplyr::full_join(penalty_df, by = c("player_id")) %>%
+  dplyr::full_join(touchdown_df, by = c("player_id")) %>%
+  dplyr::left_join(nflfastR::fast_scraper_roster(2020) %>% dplyr::select(player_id=gsis_id,player_name=full_name,recent_team=team,position)
```

Comment:
What is this roster join for? Why don't you compute name and team from the pbp data?

Reply:
team should come from defteam and not from rosters here (rosters will update when a player is traded midseason but the stat belongs to the team inside the pbp)

Reply:
Fair point. Thought for stat purposes it might be better to have the recent team. I don't have the problem with the static year in that case, so it works for me 🙈